[DataGrid] Improve grouping performance for large datasets

romgrk opened the pull request:
Closes #9197 

We should be using a double-linked list for this use-case, we care only about insertion time here. The memory pressure of constantly re-allocating the `sortedRows` array is turning this into an `O(n^something)` runtime. There probably is a more elegant way to solve this as I haven't used dlinked list in a while but this will do for now.

_edit: or maybe just a single-linked list is enough? I'll update tomorrow_

Before: https://codesandbox.io/s/cranky-pike-i1xy45
![Screenshot from 2023-06-02 11-51-33](https://github.com/mui/mui-x/assets/1423607/ec2a169a-686b-4d2e-b1c0-8db2f86daca6)


After: https://codesandbox.io/s/nervous-grothendieck-n1bwgf
![image](https://github.com/mui/mui-x/assets/1423607/b9722269-5681-4bae-a95c-2bb44cb14e43)

romgrk commented:
Is it normal that it only says 944 total rows (with a 100,000 rows dataset) when grouping is on? I've never worked with the grouping code yet, not sure if my PR is wrong or this is how it displays.

![image](https://github.com/mui/mui-x/assets/1423607/58fac340-db58-47d3-8728-4b6dcd2cfb90)

romgrk commented on packages/grid/x-data-grid-pro/src/utils/tree/sortRowTree.ts:
```diff
@@ -68,24 +151,22 @@ export const sortRowTree = (params: SortRowTreeParams) => {
       if (children?.length) {
-        const subTreeSize = insertRowListIntoSortedRows(startIndex + treeSize, children);
-        treeSize += subTreeSize;
+        const sublist = List.from(children);
+        const subnodes = sublist.nodes();
+        node.insertAfter(sublist);
+        fillList(subnodes);
```
I think that if I was iterating the list in the reverse order, then it would be possible to fill it without storing the list into a temporary `nodes` array.

cherniavskii replied:
Not 100% sure, but if we use `List.forEach()` instead of `List.nodes().forEach()`, we  don't even need recursion here:

```ts
const rootList = List.from<GridRowId>(sortedGroupedByParentRows.get(GRID_ROOT_GROUP_ID)!);

rootList.forEach((node) => {
  const children = sortedGroupedByParentRows.get(node.data);

  if (children?.length) {
    const sublist = List.from(children);
    node.insertAfter(sublist);
  }
});
```

romgrk replied:
Done

romgrk replied:
No recursion single-linked list, 100,000 rows:

![image](https://github.com/mui/mui-x/assets/1423607/3240552f-5f92-4d0f-8cb2-62e024114eb0)

romgrk replied:
There was indeed no need for recursion nor inversion. I was just very tired when I wrote this.

MBilalShafi reviewed (commented):
Impressive how massive a performance jump we can get just by changing the data structure 🎉 

> Is it normal that it only says 944 total rows (with a 100,000 rows dataset) when grouping is on? I've never worked with the grouping code yet, not sure if my PR is wrong or this is how it displays.

Yes, this count shows the root level nodes. The children of the node with `GRID_ROOT_GROUP_ID` id.

cherniavskii commented on packages/grid/x-data-grid-pro/src/utils/tree/sortRowTree.ts:
```diff
@@ -13,9 +13,92 @@ interface SortRowTreeParams {
+  data: T;
+
+  constructor(data: T, previous: null | Node<T>, next: null | Node<T>) {
+    this.previous = previous;
```
> edit: or maybe just a single-linked list is enough? I'll update tomorrow

👍

romgrk replied:
Done

cherniavskii reviewed (approved):
Wow, this is amazing! 🤯 🚀
Thank you, @romgrk! 🎉

romgrk commented:
> Impressive how massive a performance jump we can get just by changing the data structure

Yeah, memory IO is usually the bottleneck for the majority of software. I highly recommend [this PDF](https://people.freebsd.org/~lstewart/articles/cpumemory.pdf) on the subject.

romgrk commented on packages/grid/x-data-grid-pro/src/hooks/features/lazyLoader/useGridLazyLoaderPreProcessors.tsx:
```diff
@@ -19,8 +19,7 @@ export const useGridLazyLoaderPreProcessors = (
 
   const addSkeletonRows = React.useCallback<GridPipeProcessor<'hydrateRows'>>(
     (groupingParams) => {
-      const tree = { ...groupingParams.tree };
-      const rootGroup = tree[GRID_ROOT_GROUP_ID] as GridGroupNode;
```
On the example for the linked issue, this was costing 45ms for nothing.

romgrk commented:
I've cleaned up the linked list implementation & turned it into a single-linked one.

I've also added a few other small changes, mainly avoiding an object spread that wasn't used, and unwrapping object params into just raw params.

| Before | After |
| --- | --- |
| ![total-before](https://github.com/mui/mui-x/assets/1423607/138f11aa-0357-4dc7-a0da-8f67618230d9) | ![total-after](https://github.com/mui/mui-x/assets/1423607/cf41b0bc-138a-44a0-b33e-a4f9360f68cd) |

We can still do more gains but it will need changing the data model and I need some stuff from #9120 in first.
